ast/parser.go

```diff
@@ -795,43 +795,36 @@ func (p *Parser) parseElse(head *Head) *Rule {
 	}
 
 	hasIf := p.s.tok == tokens.If
+	hasLBrace := p.s.tok == tokens.LBrace
 
-	if hasIf {
-		p.scan()
-		s := p.save()
-		if expr := p.parseLiteral(); expr != nil {
-			// NOTE(sr): set literals are never false or undefined, so parsing this as
-			//  p if false else if { true }
-			//                     ^^^^^^^^ set of one element, `true`
-			// isn't valid.
-			isSetLiteral := false
-			if t, ok := expr.Terms.(*Term); ok {
-				_, isSetLiteral = t.Value.(Set)
-			}
-			// expr.Term is []*Term or Every
-			if !isSetLiteral {
-				rule.Body.Append(expr)
-				setLocRecursive(rule.Body, rule.Location)
-				return &rule
-			}
-		}
-		p.restore(s)
-	}
-
-	if p.s.tok != tokens.LBrace {
+	if !hasIf && !hasLBrace {
 		rule.Body = NewBody(NewExpr(BooleanTerm(true)))
 		setLocRecursive(rule.Body, rule.Location)
 		return &rule
 	}
 
-	p.scan()
+	if hasIf {
+		p.scan()
+	}
 
-	if rule.Body = p.parseBody(tokens.RBrace); rule.Body == nil {
+	if p.s.tok == tokens.LBrace {
+		p.scan()
+		if rule.Body = p.parseBody(tokens.RBrace); rule.Body == nil {
+			return nil
+		}
+		p.scan()
+	} else if p.s.tok != tokens.EOF {
+		expr := p.parseLiteral()
+		if expr == nil {
+			return nil
+		}
+		rule.Body.Append(expr)
+		setLocRecursive(rule.Body, rule.Location)
+	} else {
+		p.illegal("unexpected end of input")
```

Comment:
This error note will create an error message with a bit of redundancy in it. `p.illegat()` will add `unexpected eof token`to the error message, suffixed with this message; i.e. `rego_parse_error: unexpected eof token: unexpected end of input`.

Let's change this to something a bit more specific to the detected error state; e,g, `rule body expected`; which'd give us `rego_parse_error: unexpected eof token: rule body expected`

Reply:
Thank you so much for the review and the great comments, I have updated the code accordingly!